DM-38444 Add Sasquatch datastore

natelust opened the pull request:
## Checklist

- [ ] ran Jenkins
- [x] added a release note for user-visible changes to `doc/changes`

timj commented on doc/changes/DM-38444.md:
```diff
@@ -0,0 +1 @@
+Introduced a special datastore to upload metric measurements to a Sasquatch instance when a MetricMeasurementBundle is stored with a butler.put call.
```
```suggestion
Introduced a special datastore to upload metric measurements to a Sasquatch instance when a `MetricMeasurementBundle` is stored with a `butler.put()` call.
```

Also can the full name of the datastore be included so people know where to find it?

timj commented on python/lsst/daf/butler/datastores/sasquatchDatastore.py:
```diff
@@ -0,0 +1,885 @@
+                    "One or more records may have failed to upload, or only partially succeeded"
+                )
+        else:
+            log.debug("Could not put dataset type %s with Sasquatch datastore", ref.datasetType)
```
This is supposed to raise `DatasetTypeNotSupportedError`. That is the convention when a dataset fails the constraints.

timj commented on python/lsst/daf/butler/datastores/sasquatchDatastore.py:
```diff
@@ -0,0 +1,885 @@
+                self._dispatcher.dispatchRef(inMemoryDataset, ref)
+            except SasquatchDispatchPartialFailure:
+                raise DatasetPutError(
+                    "One or more records may have failed to upload, or only partially succeeded"
```
In a chained datastore this message is going to be swallowed. Do you want a log entry to be written somewhere? Should you include the sasquatch error text? Should you raise `from e`? (I'm looking forward to python 3.11 where we can add this message to the other exception)

natelust replied:
I changed around how exceptions are being handled

natelust replied:
Actually this is going to need some discussion

timj commented on python/lsst/daf/butler/datastores/sasquatchDatastore.py:
```diff
@@ -0,0 +1,885 @@
+        super().__init__(config, bridgeManager)
+
+        # Name ourselves either using an explicit name or a name
+        # derived from the (unexpanded) root
```
```suggestion
        # derived from the (unexpanded) root.
```

timj commented on python/lsst/daf/butler/datastores/sasquatchDatastore.py:
```diff
@@ -0,0 +1,885 @@
+
+        # Name ourselves either using an explicit name or a name
+        # derived from the (unexpanded) root
+        self.name = self.config.get("name", "{}@{}".format(type(self).__name__, self.config["root"]))
```
I don't see an example yaml configuration in this PR so I'm not sure what `root` means here. Are you always explicitly naming with the `name` or should the URL be used?

natelust replied:
I cribbed this from another datastore prior to fully understanding how it all worked and fit together, now that I do I think I need to re-visit that. I think the URL is reasonably unique

timj commented on python/lsst/daf/butler/datastores/sasquatchDatastore.py:
```diff
@@ -0,0 +1,885 @@
+
+        self.restProxyUrl = self.config["restProxyUrl"]
+
+        self.accessToken = self.config.get("accessToken", "na")
```
Does "na" mean something magic? Are you assuming the access token will be read from a local `sasquatchDatastore.yaml` (via `$DAF_BUTLER_CONFIG_PATH`) that will merge with this one and have the token in it? (we all agreed the token won't be in the public file didn't we?)

natelust replied:
There is no magic there. And in fact there is no "there" there. The token does nothing at this point. The token will not be public in the file, and will be specific to each user. The details are still being worked out, but it might even give the ability to scope what namespace a user is allowed to write into. I was thinking either that config path variable, or some other shell expansion if needed.

timj commented on python/lsst/daf/butler/configs/datastores/sasquatchDatastore.yaml:
```diff
@@ -0,0 +1,3 @@
+constraints:
```
This file needs to have the `cls` entry included in it. At the very least that will ensure that this reference sasquatch configuration file tells people where to look for the datastore.  Please look at `fileDatastore.yaml` as an example.

timj commented on python/lsst/daf/butler/datastores/chainedDatastore.py:
```diff
@@ -411,7 +412,9 @@ def put(self, inMemoryDataset: Any, ref: DatasetRef) -> None:
-            if constraints is not None and not constraints.isAcceptable(ref):
+            if (
+                constraints is not None and not constraints.isAcceptable(ref)
+            ) or not datastore.constraints.isAcceptable(ref):
```
This might be a nice short circuit but we do still expect a child datastore to raise `DatasetTypeNotSupportedError` so that if someone uses the datastore directly they get a consistent answer.

timj commented on python/lsst/daf/butler/datastores/chainedDatastore.py:
```diff
@@ -424,7 +427,7 @@ def put(self, inMemoryDataset: Any, ref: DatasetRef) -> None:
                 if not datastore.isEphemeral:
                     isPermanent = True
-            except DatasetTypeNotSupportedError:
+            except (DatasetTypeNotSupportedError, DatasetPutError):
```
We have to be a bit careful here because `DatasetPutError` is documented as being the exception that is raised when a `put` fails and so one would expect that we should change `FileDatastore` to also use it. If we do that though the semantics of `ChainedDatastore.put()` change completely. Previous to this, a failure of any kind from a child datastore that was not due to constraints failure caused the entire `put` to fail. Maybe that was wrong all along and that we should have caught any failure in put and only raised later on if `nsuccess` was 0 (and really if none worked we'd probably want to cache the most recent error and raise that if `nsuccess` was 0 and possibly include the other errors if there was more than one failure). If we are catching multiple exception types we can't raise `DatasetTypeNotSupportedError` on the line below.

If we aren't raising it at all we probably should be issuing some kind of log entry because otherwise it implies that we can be in a situation where nothing gets put to sasquatch and we can only tell by noticing that everything is missing.

natelust replied:
I am uploading a change, lets talk about it when you see it.

timj commented on python/lsst/daf/butler/datastores/sasquatchDatastore.py:
```diff
@@ -0,0 +1,885 @@
+    """The namespace in Sasquatch in which to write the uploaded metrics"""
+
+    def __post_init__(self) -> None:
+        parsedUrl = urllib.parse.urlparse(self.url)
```
You could use `ResourcePath` here.

timj commented on python/lsst/daf/butler/datastores/sasquatchDatastore.py:
```diff
@@ -0,0 +1,885 @@
+            try:
+                self._dispatcher.dispatchRef(inMemoryDataset, ref)
+            except SasquatchDispatchPartialFailure:
+                raise DatasetPutError(
```
If you get a sasquatch error that is not due to a partial failure, does that end up being propagated up directly and won't be a `DatasetPutError`?

natelust replied:
I removed DatasetPutError, but  this needs to be part of the larger discussion on Exceptions

timj commented on python/lsst/daf/butler/datastores/sasquatchDatastore.py:
```diff
@@ -0,0 +1,885 @@
+    pass
+
+
+def _tag2VersionTime(productStr: str) -> tuple[str, float]:
```
This is a new file. You can use `_tag_to_version_time` and use underscores in the variables.

natelust replied:
Personally, I prefer camel case, though part of that may be the muscle memory from the years of doing it.

timj commented on python/lsst/daf/butler/datastores/sasquatchDatastore.py:
```diff
@@ -0,0 +1,885 @@
+    """Determine versions and dates from the string returned from
+    getEnvironmentPackages.
+
+    The genEnvironmentPackages function returns the setup version associated
```
```suggestion
    The `~lsst.utils.packages.getEnvironmentPackages` function returns the setup version associated
```
(will need to wrap the sentence)

timj commented on python/lsst/daf/butler/datastores/sasquatchDatastore.py:
```diff
@@ -0,0 +1,885 @@
+    -------
+    result : `tuple` of `str`, `datetime.datetime`
+        The first `str` is the version of the package, and the second is the
+        datetime object associates with that released version.
```
```suggestion
        datetime object associated with that released version.
```

timj commented on python/lsst/daf/butler/datastores/sasquatchDatastore.py:
```diff
@@ -0,0 +1,885 @@
+            case ("w", year, week):
+                iyear = int(year)
+                iweek = int(week)
+                if (diff := iweek - 52) > 0:
```
Maybe add a comment explaining that we can go all the way to `w_2022_53`. (which was made on 2022-12-29 but refers to 2022-12-28 since it seems like there was a delay in getting it out).

natelust replied:
Actually I am not sure what to do about this yet, it seems like no matter what we do, some times data is going to be overwritten. I think I might have to translate the extra weeks to some time after week 52 or something

timj replied:
Maybe the simplest hack is to declare that if you see week 53 then the date is 202X-12-31T23:59 or something. It should never roll over to the next year.

natelust replied:
Agreed, except if we end up with 54. I think I am going to increment the date. That will give us 4 extra "weeklies" in case.

timj replied:
You don't have to worry about 54 -- that is impossible. 53 is a valid week -- `fromisocalendar` will accept it, but only for years that actually have 53 weeks.

timj commented on python/lsst/daf/butler/datastores/sasquatchDatastore.py:
```diff
@@ -0,0 +1,885 @@
+                    iyear += (diff // 52) + 1
+                    iweek = diff % 52
+                # Use 3 as the day because we cut releases on Wednesday.
+                dt = datetime.datetime.fromisocalendar(iyear, iweek, 3)
```
I am a tad concerned with our weekly tag calculations. 2020 had 53 weeks but 2022 does not (according to iso) so how do we end up with w_2022_53 which `fromisocalendar` correctly complains about -- it should be impossible regardless of which day of the week we calculated it at the end of the year.

@athornton or @ktlim do you know where we calculate the tag?

If I plug `w_2022_53` into this algorithm I get back 2023-01-04 and not 2022-12-28 (obviously we get the same date as `w_2023_01`. Additionally, the valid ISO tag of `w_2020_53` gives the invalid 2021-01-06 when it should be reporting 2020-12-30. Also, w.2022.01 was issued on 2021-12-30 (and not 2022-01-05) but I'm not sure if that is because somehow the weekly is calculated for the end of the week or something?

It seems to me that the options here are:

* Look at the code that makes the tags and try to reverse engineer it here.
* Abandon the whole idea and generate an integer 202253 (but then hope that it doesn't have a big gap when it displays 202301)
* Change sconsUtils so that we store the git tag date in the `version.py` and then we use that date directly.

timj replied:
@ktlim tells me that the weekly tag is calculated here: https://github.com/lsst-dm/jenkins-dm-jobs/blob/main/pipelines/release/weekly_release_cron.groovy#L27-L31

and somehow that is generating week numbers that are not ISO compliant.

natelust replied:
I have modified the date algorithm and tested it against the dates as reported on `https://eups.lsst.codes/stack/src/tags/`

timj commented on python/lsst/daf/butler/datastores/sasquatchDatastore.py:
```diff
@@ -0,0 +1,885 @@
+    version = productStr.split()[0]
+    tags: str = re.findall("[(](.*)[)]", productStr)[0]
+    for tag in tags.split():
+        match tag.split("_"):
```
Can we also try to support `w.2022.12` style git tags as well as EUPS tags?

natelust replied:
Can you imagine a case where we would have them mixed?

natelust replied:
I put in a work around.

timj replied:
You are right that in PyPI versions you would get versions like `25.2023.1700` for the weekly.

timj commented on python/lsst/daf/butler/datastores/sasquatchDatastore.py:
```diff
@@ -0,0 +1,885 @@
+        headers = {"content-type": "application/json"}
+        r = requests.get(f"{self.url}/v3/clusters", headers=headers)
+
+        if r.status_code == 200:
```
```suggestion
        if r.status_code == requests.codes.ok:
```

timj commented on python/lsst/daf/butler/datastores/sasquatchDatastore.py:
```diff
@@ -0,0 +1,885 @@
+        if r.status_code == 201:
+            log.debug("Created topic %s.%s", self.namespace, topic_name)
+            return True
+        elif r.status_code == 400:
```
```suggestion
        elif r.status_code == requests.codes.bad_request:
```

timj commented on python/lsst/daf/butler/datastores/sasquatchDatastore.py:
```diff
@@ -0,0 +1,885 @@
+            f"{self.url}/v3/clusters/{self.clusterId}/topics", json=topic_config, headers=headers
+        )
+
+        if r.status_code == 201:
```
```suggestion
        if r.status_code == requests.codes.created:
```

timj commented on python/lsst/daf/butler/datastores/sasquatchDatastore.py:
```diff
@@ -0,0 +1,885 @@
+        Parameters
+        ----------
+        value : `Any`
+            Any python parameter
```
```suggestion
            Any python parameter.
```

timj commented on python/lsst/daf/butler/datastores/sasquatchDatastore.py:
```diff
@@ -0,0 +1,885 @@
+                    return {}
+                if len(tmp) > 1:
+                    log.error(
+                        "Sequence contains mixed types: %s, must be heterogeneous for avro conversion "
```
Do you mean homogeneous here?

timj commented on python/lsst/daf/butler/datastores/sasquatchDatastore.py:
```diff
@@ -0,0 +1,885 @@
+            case Sequence():
+                tmp = set()
+                for item in value:
+                    tmp.add(self._python2Avro(item)["type"])
```
```python
tmp = {self._python2Avro(item)["type"] for item in value}
```
?

timj commented on python/lsst/daf/butler/datastores/sasquatchDatastore.py:
```diff
@@ -0,0 +1,885 @@
+                return {}
+
+    def _handleReferencePackage(self, meta: MutableMapping, bundle: MetricMeasurementBundle) -> None:
+        """Check to see if there is a reference package, if there is determine
```
This is too long as a summary line.

```python
"""Check to see if there is a reference package.

If there is a reference package determine the datetime associated 
with this reference package. Save the package...
"""
```

timj commented on python/lsst/daf/butler/datastores/sasquatchDatastore.py:
```diff
@@ -0,0 +1,885 @@
+            The bundled metrics
+        """
+        ref_package, package_version, package_timestamp = "", "", 0.0
+        if hasattr(bundle, "reference_package") and bundle.reference_package:
```
```python
if ref_package := getattr(bundle, "reference_package", ""):
```
? and then no need for the line below.

timj commented on python/lsst/daf/butler/datastores/sasquatchDatastore.py:
```diff
@@ -0,0 +1,885 @@
+            A mapping which corresponds to fields which should be encoded in
+            all records.
+        bundle : `MetricMeasurementBundle`
+            The bundled metrics
```
`run` parameter is not documented.

timj commented on python/lsst/daf/butler/datastores/sasquatchDatastore.py:
```diff
@@ -0,0 +1,885 @@
+                case "current_timestamp":
+                    timeStamp = datetime.datetime.now()
+                    meta["timestamp"] = timeStamp.timestamp()
+                case "dataset_timestamp":
```
This is the one where we get the date from the exposure/visit record?

natelust replied:
Yes, and working all that out is being saved for a different ticket.

timj commented on python/lsst/daf/butler/datastores/sasquatchDatastore.py:
```diff
@@ -0,0 +1,885 @@
+        """
+        identifier: str = "Generic"
+        if hasattr(bundle, "datasetIdentifier") and bundle.datasetIdentifier is not None:
+            identifier = bundle.datasetIdentifier
```
It's a shame this doesn't work:
```python
identifier = getattr(bundle, "datasetIdentifier", "Generic")
```
since it would all be in one line, but then you would need to make it a bit more annoying to handle the `is not None`

natelust replied:
I could go with `identifier: str = x if (x := getattr(bundle, 'dataset_identifier', None)) is not None else "Generic"`

timj commented on python/lsst/daf/butler/datastores/sasquatchDatastore.py:
```diff
@@ -0,0 +1,885 @@
+        Together these values (in addition to the timestamp and topic) should
+        uniquely identify an upload to the Sasquatch system.
+
+        Parameters
```
Missing half the parameters.

timj commented on python/lsst/daf/butler/datastores/sasquatchDatastore.py:
```diff
@@ -0,0 +1,885 @@
+            identifier = bundle.datasetIdentifier
+
+        # override with any user supplied identifer
+        if datasetIdentifier is not None:
```
If this always overrides the bundle identifier should we reverse the checks?

```python
if datasetIdentifier is not None:
    identifer = datasetIdentifier
elif hasattr(bundle, "datasetIdentifier") and bundle.datasetIdentifier is not None:
    identifier = bundle.datasetIdentifier
else:
    identifier = "Generic"
```
?

timj commented on python/lsst/daf/butler/datastores/sasquatchDatastore.py:
```diff
@@ -0,0 +1,885 @@
+        Parameters
+        ----------
+        bundle : `~lsst.analysis.tools.MetricMeasurementBundle`
+            The bundle containing metric values to upload
```
```suggestion
            The bundle containing metric values to upload.
```

timj commented on python/lsst/daf/butler/datastores/sasquatchDatastore.py:
```diff
@@ -0,0 +1,885 @@
+            database. If this value is None, timestamp will be set by the run
+            time or current time.
+        id : `UUID` or `None`
+            The UUID of the MetricMeasurementBundle within the butler. If None,
```
```suggestion
            The UUID of the `MetricMeasurementBundle` within the butler. If `None`,
```

timj commented on python/lsst/daf/butler/datastores/sasquatchDatastore.py:
```diff
@@ -0,0 +1,885 @@
+            is also stored in the butler, this should be the butler run
+            collection the bundle is stored in the butler.
+        datasetType : `str`
+            The dataset type name associated with this MetricMeasurementBundle
```
```suggestion
            The dataset type name associated with this `MetricMeasurementBundle`.
```

timj commented on python/lsst/daf/butler/datastores/sasquatchDatastore.py:
```diff
@@ -0,0 +1,885 @@
+        Returns
+        -------
+        result : `Mapping` of `str` to `list`
+            A mapping of metric name of list of metric measurement records
```
```suggestion
            A mapping of metric name of list of metric measurement records.
```

timj commented on python/lsst/daf/butler/datastores/sasquatchDatastore.py:
```diff
@@ -0,0 +1,885 @@
+        result : `Mapping` of `str` to `list`
+            A mapping of metric name of list of metric measurement records
+        status : `bool`
+            A status bool indicating if some records had to be skipped due to
```
```suggestion
            A status boolean indicating if some records had to be skipped due to
```

timj commented on python/lsst/daf/butler/datastores/sasquatchDatastore.py:
```diff
@@ -0,0 +1,885 @@
+            # associated with metric
+            metricRecordList = metricRecords.setdefault(metric, list())
+
+            record: dict[str, Any] = dict()
```
Is it more efficient to do this as:

```python
record = meta.copy()
```
rather than using `update()` immediately?

timj commented on python/lsst/daf/butler/datastores/sasquatchDatastore.py:
```diff
@@ -0,0 +1,885 @@
+            # loop over each metric measurement within the metric
+            for measurement in measurements:
+                # need to extract any tags, package info, etc
+                notes = dict(measurement.notes.items())
```
```python
notes = measurement.notes.copy()
```
? But I'm not sure why `notes` exists at all because it's only used in the line below.

natelust replied:
old iteration of the code, i'll change

timj commented on python/lsst/daf/butler/datastores/sasquatchDatastore.py:
```diff
@@ -0,0 +1,885 @@
+                notes = dict(measurement.notes.items())
+                record["tags"] = notes.get("metric_tags", list())
+
+                # Missing values are replaced by 0 in sasquatch, see RFC-763
```
```suggestion
                # Missing values are replaced by 0 in sasquatch, see RFC-763.
```

timj commented on python/lsst/daf/butler/datastores/sasquatchDatastore.py:
```diff
@@ -0,0 +1,885 @@
+                # Missing values are replaced by 0 in sasquatch, see RFC-763
+                name = ""
+                value = 0.0
+                match measurement.json:
```
`.json` returns a `dict` and not a `str`? Pydantic has `.dict()` return a dict and `.json()` return the json string. It seems like this `json` property doesn't return JSON and I'm now confused.

natelust replied:
yes, a `lsst.verify.Measurement` property `.json` returns a python dict that is compatible with `json.dumps`

timj commented on python/lsst/daf/butler/datastores/sasquatchDatastore.py:
```diff
@@ -0,0 +1,885 @@
+                    case {"metric": name, "value": None}:
+                        pass
+                    case {"metric": name, "value": value}:
+                        if value != value:
```
Should we be using the more explicit:
```python
if math.isnan(value):
```
rather than being this abstract?

natelust replied:
There was some edge case where something was not considering something nan, mixing type implementations or something like that, I cant remember what now though. We can switch to `math` and see if the edge case comes up again, I dont have strong feelings on it, outside this is exactly the definition of nan by the IEEE, and so I think it is a reasonable choice

timj replied:
I do feel that `math.isnan` is more explicit and it is implemented in C so it shouldn't be slower than what you have. If you still want to do the `x != x` code please add a comment -- not every python programmer will understand what you mean there.

timj commented on python/lsst/daf/butler/datastores/sasquatchDatastore.py:
```diff
@@ -0,0 +1,885 @@
+        # Look at each of the metrics in the bundle (name, values)
+        for metric, measurements in bundle.items():
+            # create a list which will contain the records for each measurement
+            # associated with metric
```
```suggestion
            # associated with metric.
```

timj commented on python/lsst/daf/butler/datastores/sasquatchDatastore.py:
```diff
@@ -0,0 +1,885 @@
+
+        # Look at each of the metrics in the bundle (name, values)
+        for metric, measurements in bundle.items():
+            # create a list which will contain the records for each measurement
```
```suggestion
            # Create a list which will contain the records for each measurement
```

timj commented on python/lsst/daf/butler/datastores/sasquatchDatastore.py:
```diff
@@ -0,0 +1,885 @@
+        Parameters
+        ----------
+        bundle : `~lsst.analysis.tools.MetricMeasurementBundle`
+            The bundle containing metric values to upload
```
```suggestion
            The bundle containing metric values to upload.
```

timj commented on python/lsst/daf/butler/datastores/sasquatchDatastore.py:
```diff
@@ -0,0 +1,885 @@
+            collection the bundle is stored in the butler. This will be used
+            in generating uniqueness constraints in Sasquatch.
+        datasetType : `str`
+            The dataset type name associated with this MetricMeasurementBundle
```
```suggestion
            The dataset type name associated with this `MetricMeasurementBundle`.
```

timj commented on python/lsst/daf/butler/datastores/sasquatchDatastore.py:
```diff
@@ -0,0 +1,885 @@
+            database. If this value is None, timestamp will be set by the run
+            time or current time.
+        id : `UUID` or `None`
+            The UUID of the MetricMeasurementBundle within the butler. If None,
```
```suggestion
            The UUID of the `MetricMeasurementBundle` within the Butler. If `None`,
```

timj commented on python/lsst/daf/butler/datastores/sasquatchDatastore.py:
```diff
@@ -0,0 +1,885 @@
+            Sasquatch will have a unique value.
+        datasetIdentifier : `str` or None
+            A string which will be used in creating unique identifier tags. If
+            None, a default value will be inserted.
```
```suggestion
            `None`, a default value will be inserted.
```

timj commented on python/lsst/daf/butler/datastores/sasquatchDatastore.py:
```diff
@@ -0,0 +1,885 @@
+
+            r = requests.post(f"{self.url}/topics/{self.namespace}.{metric}", json=data, headers=headers)
+
+            if r.status_code == 200:
```
```suggestion
            if r.status_code == requests.codes.ok:
```

timj commented on python/lsst/daf/butler/datastores/sasquatchDatastore.py:
```diff
@@ -0,0 +1,885 @@
+        Parameters
+        ----------
+        bundle : `~lsst.analysis.tools.MetricMeasurementBundle`
+            The bundle containing metric values to upload
```
```suggestion
            The bundle containing metric values to upload.
```

timj commented on python/lsst/daf/butler/datastores/sasquatchDatastore.py:
```diff
@@ -0,0 +1,885 @@
+            The timestamp to be associated with the measurements in the ingress
+            database. If this value is None, timestamp will be set by the run
+            time or current time.
+        extraFields: `Mapping` or None
```
```suggestion
        extraFields: `Mapping` or `None`
```

timj commented on python/lsst/daf/butler/datastores/sasquatchDatastore.py:
```diff
@@ -0,0 +1,885 @@
+        extraFields: `Mapping` or None
+            Extra mapping keys and values that will be added as fields to the
+            dispatched record if not None.
+        datasetIdentifier : `str` or None
```
```suggestion
        datasetIdentifier : `str` or `None`
```

timj commented on python/lsst/daf/butler/datastores/sasquatchDatastore.py:
```diff
@@ -0,0 +1,885 @@
+        serializedRef = ref.to_simple()
+        id = serializedRef.id
+        if serializedRef.run is None:
+            run = ""
```
```suggestion
            run = "<unknown>"
```
? Since this should be impossible (and soon mypy will tell you this).

timj commented on python/lsst/daf/butler/datastores/sasquatchDatastore.py:
```diff
@@ -0,0 +1,885 @@
+        else:
+            run = serializedRef.run
+        dstype = serializedRef.datasetType
+        datasetType = dstype.name if dstype is not None else ""
```
Is it possible for `SerializedDatasetRef` to have undefined dataset type and dataId?

natelust replied:
According to the typing yes, but this handles those cases.

timj commented on python/lsst/daf/butler/datastores/sasquatchDatastore.py:
```diff
@@ -0,0 +1,885 @@
+class SasquatchDatastore(GenericBaseDatastore):
+    """Basic Datastore for writing to an in Sasquatch instance.
+
+    This Datastore is write only, meaning that it can dispatch data to a
```
```suggestion
    This Datastore is currently write only, meaning that it can dispatch data to a
```

timj commented on python/lsst/daf/butler/datastores/sasquatchDatastore.py:
```diff
@@ -0,0 +1,885 @@
+    """Basic Datastore for writing to an in Sasquatch instance.
+
+    This Datastore is write only, meaning that it can dispatch data to a
+    sasquatch instance, but at the present can not be used to retrieve values.
```
```suggestion
    Sasquatch instance, but at the present can not be used to retrieve values.
```

timj commented on python/lsst/daf/butler/datastores/sasquatchDatastore.py:
```diff
@@ -0,0 +1,885 @@
+        return
+
+    def validateKey(self, lookupKey: LookupKey, entity: DatasetRef | DatasetType | StorageClass) -> None:
+        # Docstring is inherited from base class
```
```suggestion
        # Docstring is inherited from base class.
```

timj commented on python/lsst/daf/butler/datastores/sasquatchDatastore.py:
```diff
@@ -0,0 +1,885 @@
+        return
+
+    def getLookupKeys(self) -> set[LookupKey]:
+        # Docstring is inherited from base class
```
```suggestion
        # Docstring is inherited from base class.
```

timj commented on python/lsst/daf/butler/datastores/sasquatchDatastore.py:
```diff
@@ -0,0 +1,885 @@
+    def export_records(self, refs: Iterable[DatasetIdRef]) -> Mapping[str, DatastoreRecordData]:
+        # Docstring inherited from the base class.
+
+        # In-memory Datastore records cannot be exported or imported
```
```suggestion
        # Sasquatch Datastore records cannot be exported or imported.
```

timj reviewed (changes requested):
Thanks for all this. I've made lots of comments and will take another look.

There needs to be some text added to `doc/lsst.daf.butler/datastores.rst` explaining the datastore and, importantly, the configuration options.

timj commented:
An additional point I've just noticed, you need to make sure that `self.bridge.insert(ref)` is called at some point during your put.  This populates the `dataset_locations` table to indicate that the ref is associated with the named datastore. In theory this can be used externally to find out which datastores know about a given dataset. If you do this you will also have to include a trash method and emptyTrash that moves the row out of dataset_locations even if sasquatch still has the dataset (which is also what the `datastore.forget` method can do).

natelust commented:
Do we want to insert a record into a table about who knows about this data? I would prefer not to at this moment because a) we cant fetch anything from Sasquatch anyway b) we cant do anything about the fact that someone might delete the record on the sasquatch side. Have a table that is wrong seems worse than not having a table entry c) the worst that will happen if someone runs something again is the same value will just be uploaded again.

timj commented:
Regarding recording that the datastore has accepted it, I'm fine with us punting for now but we do need to address it at some point.

timj commented on python/lsst/daf/butler/core/exceptions.py:
```diff
@@ -37,3 +37,13 @@ class ValidationError(RuntimeError):
     pass
+
+
+class DatasetPutError(RuntimeError):
```
I don't think this is used anywhere now so can be deleted.

timj reviewed (approved):
Thanks for the clean ups.

natelust committed:
Small changes in Chained Datastore

Some minor changes were made in chained datastore to support the
read only SasquatchDatastore going into analysis_tools. A note
of the datastore was added to the datastore document.
